Add support for `import defer` proposal

nicolo-ribaudo opened the pull request:
<!--
Thank you for submitting a pull request!

Please verify that:
* [ ] There is an associated issue in the `Backlog` milestone (**required**)
* [ ] Code is up-to-date with the `main` branch
* [ ] You've successfully run `hereby runtests` locally
* [ ] There are new or updated unit tests validating the change

Refer to CONTRIBUTING.MD for more details.
  https://github.com/Microsoft/TypeScript/blob/main/CONTRIBUTING.md

** Please don't send typo fixes! **
Please don't send a PR solely for the purpose of fixing a typo, unless that
typo truly hurts understanding of the text. Each PR represents work for the
maintainers, and that work should provide commensurate value.

If you're interested in sending a PR, the issue tracker has many issues marked `help wanted`.
-->

This proposal is currently at Stage 3.

This PR only needs to add _parsing_ support for the proposal:
- it does not support downleveling it, as it's not possible to transpile it to older ESM versions
  - it would be possible add support for transpiling to CommonJS, if needed.
- it does not introduce any type-checking changes, as the object created by `import defer * as ns` is meant to "look like" the one created by `import * as ns`

As you'll see from the code, whether a module is deferred or not is not a boolean but an enum. That's because of the `import source` proposal: there is no PR for it yet, but once it will be implemented it should be done by adding one more possible "import phase" to this enum. A question I have is if that should be an enum, or just an union of `undefined | DeferKeyword` (which will one day become `undefined | DeferKeyword | SourceKeyword`).

Fixes https://github.com/microsoft/TypeScript/issues/59391

_This patch was originally written by @ryzokuken_

jakebailey commented on src/compiler/factory/nodeFactory.ts:
```diff
@@ -4723,11 +4724,12 @@ export function createNodeFactory(flags: NodeFactoryFlags, baseFactory: BaseNode
 
     // @api
-    function createImportClause(isTypeOnly: boolean, name: Identifier | undefined, namedBindings: NamedImportBindings | undefined): ImportClause {
+    function createImportClause(isTypeOnly: boolean, name: Identifier | undefined, namedBindings: NamedImportBindings | undefined, phase: ImportPhase): ImportClause {
```
Preemptively noting that this is a breaking API change, and would require a deprecation helper in the deprecations project to tack onto our public API something that will set a default for the new parameter.

jakebailey replied:
It may also be the case that `phase` needs to go after `isTypeOnly` since AST node builder parameters and properties are intended to be in source order.

rbuckton replied:
I don't think we want to mix `type` imports with `defer` or `source`. Something like `import type source from "foo"` doesn't make sense if all `source` imports will have the same type, and `import type defer * as foo from "foo"` would essentially be the same as `import type * as foo from "foo"`.

If we consider `type`, `defer`, and `source` to be mutually exclusive, then I would suggest we replace the `isTypeOnly` parameter with something like `importModifier: SyntaxKind.TypeKeyword | SyntaxKind.DeferKeyword | SyntaxKind.SourceKeyword | boolean | undefined` and have `ImportClause` be: 

```ts
export interface ImportClause extends NamedDeclaration {
    readonly kind: SyntaxKind.ImportClause;
    readonly parent: ImportDeclaration | JSDocImportTag;
    /** @deprecated */
    readonly isTypeOnly: boolean;
    readonly importModifier: SyntaxKind.TypeKeyword | SyntaxKind.DeferKeyword | SyntaxKind.SourceKeyword | undefined;
    readonly name?: Identifier; // Default binding
    readonly namedBindings?: NamedImportBindings;
}
```

We can set `isTypeOnly` to `true` if `importModifier` is `SyntaxKind.TypeKeyword`.

DanielRosenwasser replied:
I think the issue with that scheme is that it's a little weirder to capture multiple modifiers being used in the cases of error recovery.

nicolo-ribaudo replied:
I was originally going with something very similar to Ron's suggestion, and consider "type" just as another phase. I ended up not doing it because of the AST breaking change, but if just keeping the old property for backwards compat is ok then I'd go for it.

> I think the issue with that scheme is that it's a little weirder to capture multiple modifiers being used in the cases of error recovery.

Error recovery is going to be tricky anyway, because each modifier is a valid identifier so things like `import type defer` are a potentially valid start of an import declaration. I wonder how likely it is for people to accidentally write two modifiers in an import.

DanielRosenwasser commented:
> I would expect that these need to be modifiers so that they are nodes that can be walked, since people can stick comments between them and so on...

Modifiers would precede the `import` here, so it's a bit more iffy than that. But I guess the `export @decorator class C {}` thing has already made things iffy between decorators/modifiers, and @rbuckton recalls how we handled this.

But this is close syntactically to `import type`, which just gets an `isTypeOnly` property slapped on, and hasn't caused issues. When we do need to hydrate nodes in the language service, we'll end up with a synthesized `defer` node. So from a technical standpoint, I think it's doesn't cause issues, but it might not be the best represenation long-term.

rbuckton commented:
> But this is close syntactically to `import type`, which just gets an `isTypeOnly` property slapped on, and hasn't caused issues.

We have `emitTokenWithComment` to handle synthetic punctuation and comments, so that's fine. IMO we should just use `SyntaxKind` instead of a separate `enum`, as I suggested [above](https://github.com/microsoft/TypeScript/pull/60757#discussion_r1884548891).

rbuckton commented:
That said, `ImportClause` is parsed after the `import` keyword, so it could have `modifiers` and we could just make `type` a modifier with `isTypeOnly` existing purely for back-compat purposes. That would require a lot of additional complexity in how we handle modifiers, however, as we'd have to have grammar checks to disallow `type`, `defer`, and `source` as modifiers for most elements.

Jack-Works commented on src/compiler/emitter.ts:
```diff
@@ -3689,6 +3690,10 @@ export function createPrinter(printerOptions: PrinterOptions = {}, handlers: Pri
             emitTokenWithComment(SyntaxKind.TypeKeyword, node.pos, writeKeyword, node);
             writeSpace();
         }
+        else if (node.phase !== ImportPhase.Evaluation) {
```
```suggestion
        else if (node.phase === ImportPhase.Defer) {

```

nicolo-ribaudo commented:
I updated the PR to go with the suggestion in https://github.com/microsoft/TypeScript/pull/60757#discussion_r1884548891, which is what I found the cleanest. Happy to do differently if needed. The API change is now backwards compatible. Probably I should do the same change for `export`s? Even though `export defer` and `export source` don't exist, so it's only for `type`.

I also added a test showing that comments are properly preserved.

nicolo-ribaudo commented on src/compiler/emitter.ts:
```diff
@@ -3685,7 +3685,11 @@ export function createPrinter(printerOptions: PrinterOptions = {}, handlers: Pri
+            emitTokenWithComment(node.phaseModifier, node.pos, writeKeyword, node);
+            writeSpace();
+        }
+        else if (node.isTypeOnly) {
```
I have this in case somebody passes in an object from an old TS version. Is this relevant for TS' API contract, or would removing it not be considered a breaking change?

rbuckton replied:
We don't support mixing TS versions between a given `Node` and the emitter, so that shouldn't be a concern. Inside the compiler we should avoid depending on deprecated functionality.

nicolo-ribaudo commented:
Marking as draft until I implement `import.defer()`, which has been confirmed to remain part of the proposal.

nicolo-ribaudo commented:
Oh well, that was simpler than expected. Given that `import(...)` is parsed as a `CallExpression` whose `.expression` is `import`, I'm parsing `import.defer(...)` as a `CallExpression` whose `.expression` is `MetaProperty(import, defer)` (+ some validation to make sure that it's not used as a standalone meta property, but only in the call syntax).

rbuckton commented on src/compiler/checker.ts:
```diff
@@ -37642,7 +37650,7 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
         const file = getSourceFileOfNode(node);
-        Debug.assert(!!(file.flags & NodeFlags.PossiblyContainsImportMeta), "Containing file is missing import meta node flag.");
+        Debug.assert(node.name.escapedText === "defer" || !!(file.flags & NodeFlags.PossiblyContainsImportMeta), "Containing file is missing import meta node flag.");
         return node.name.escapedText === "meta" ? getGlobalImportMetaType() : errorType;
```
Should `import.defer` be returning `errorType` here? Should we also report an error here for `import.defer` outside of a call?

rbuckton commented on src/compiler/checker.ts:
```diff
@@ -52604,8 +52612,11 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
-                    return grammarErrorOnNode(node.name, Diagnostics._0_is_not_a_valid_meta_property_for_keyword_1_Did_you_mean_2, unescapeLeadingUnderscores(node.name.escapedText), tokenToString(node.keywordToken), "meta");
+                if (escapedText !== "meta" && escapedText !== "defer") {
+                    const suggestion = node.parent.kind === SyntaxKind.CallExpression && (node.parent as CallExpression).expression === node
+                        ? "meta' or 'defer"
```
This will not be properly localized due to the use of the english `or`. We should have a separate diagnostic message for this case.

rbuckton commented on src/compiler/checker.ts:
```diff
@@ -52604,8 +52612,11 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
-                if (escapedText !== "meta") {
-                    return grammarErrorOnNode(node.name, Diagnostics._0_is_not_a_valid_meta_property_for_keyword_1_Did_you_mean_2, unescapeLeadingUnderscores(node.name.escapedText), tokenToString(node.keywordToken), "meta");
+                if (escapedText !== "meta" && escapedText !== "defer") {
+                    const suggestion = node.parent.kind === SyntaxKind.CallExpression && (node.parent as CallExpression).expression === node
```
```suggestion
                    const suggestion = isCallExpression(node.parent) && node.parent.expression === node
```

rbuckton commented on src/compiler/checker.ts:
```diff
@@ -52891,6 +52905,11 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
             return grammarErrorOnNode(node, Diagnostics.ESM_syntax_is_not_allowed_in_a_CommonJS_module_when_verbatimModuleSyntax_is_enabled);
         }
 
+        if (node.expression.kind === SyntaxKind.MetaProperty) {
+            if (moduleKind !== ModuleKind.ESNext) {
```
These conditions can be combined. Wouldn't we also want to support `NodeNext`?

rbuckton commented on src/compiler/parser.ts:
```diff
@@ -5938,7 +5939,18 @@ namespace Parser {
+                        sourceFlags |= NodeFlags.PossiblyContainsDynamicImport;
+                    }
+                    else {
+                        parseErrorAt(pos, getNodePos(), Diagnostics.import_defer_must_be_followed_by);
```
Normally we would just issue a `'(' expected` diagnostic. We also may want to treat this as a grammar error (in the checker) rather than a parse error so as not to turn off other semantic checks in the file.

rbuckton commented on src/compiler/diagnosticMessages.json:
```diff
@@ -8433,5 +8433,21 @@
         "category": "Error",
         "code": 18057
+    },
+    "Default imports aren't allowed for deferred imports.": {
+        "category": "Error",
+        "code": 18058
+    },
+    "Named imports aren't allowed for deferred imports.": {
+        "category": "Error",
+        "code": 18059
+    },
```
We mostly use "are not" (38 uses) over "aren't" (1 use) in diagnostic messages, much like we prefer "is not" (99 uses) over "isn't" (1 use).

```suggestion
    "Default imports are not allowed in a deferred import.": {
        "category": "Error",
        "code": 18058
    },
    "Named imports are not allowed in a deferred import.": {
        "category": "Error",
        "code": 18059
    },
```

rbuckton commented on src/compiler/checker.ts:
```diff
@@ -52870,6 +52881,9 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
         if (node.isTypeOnly && node.namedBindings?.kind === SyntaxKind.NamedImports) {
             return checkGrammarNamedImportsOrExports(node.namedBindings);
         }
+        if (node.phaseModifier === SyntaxKind.DeferKeyword && moduleKind !== ModuleKind.ESNext) {
```
Wouldn't we also want to support `NodeNext` so that it can be used in NodeJS?

rbuckton commented on src/compiler/parser.ts:
```diff
@@ -8480,11 +8502,19 @@ namespace Parser {
+            }
+            else {
+                if (phaseModifier === SyntaxKind.DeferKeyword) {
+                    parseErrorAtCurrentToken(Diagnostics.Named_imports_aren_t_allowed_for_deferred_imports);
```
I think it would make more sense to treat this as a grammar error in the checker.

rbuckton commented on src/compiler/parser.ts:
```diff
@@ -8365,21 +8379,29 @@ namespace Parser {
+            phaseModifier = SyntaxKind.DeferKeyword;
+            identifier = undefined;
+            if (isIdentifier()) {
+                parseErrorAtCurrentToken(Diagnostics.Default_imports_aren_t_allowed_for_deferred_imports);
```
I think it would make more sense to treat this as a grammar error in the checker

rbuckton commented on src/compiler/types.ts:
```diff
@@ -3711,11 +3718,14 @@ export type NamedExportBindings =
     readonly kind: SyntaxKind.ImportClause;
     readonly parent: ImportDeclaration | JSDocImportTag;
-    readonly isTypeOnly: boolean;
+    /** @deprecated */ readonly isTypeOnly: boolean;
```
```suggestion
    /** @deprecated Use `phaseModifier` instead. */
    readonly isTypeOnly: boolean;
```

rbuckton commented on src/compiler/types.ts:
```diff
@@ -3711,11 +3718,14 @@ export type NamedExportBindings =
     readonly namedBindings?: NamedImportBindings;
 }
 
+export type ImportPhaseModifier = SyntaxKind.TypeKeyword | SyntaxKind.DeferKeyword;
```
Types ending in `Modifier` generally refer to an instantiation of the `ModifierToken` type (or are part of the `Modifier` union of those types). I'd prefer this be renamed to `ImportPhaseModifierSyntaxKind`, much like we do for `ModifierSyntaxKind`, `KeywordTypeSyntaxKind`, etc.

nicolo-ribaudo commented on tests/cases/conformance/importDefer/dynamicImportDefer.ts:
```diff
@@ -0,0 +1,14 @@
+  ns.foo();
+});
+
+import("./a.js"); // TODO: Without this the import.defer cannot resolve ./a
```
I'm going to debug this next week, but if you have any pointer as to why I would be getting an error about `./a.js` not being resolved in `import.defer(...)` without this I'd appreciate it :)

fireairforce commented:
Sorry for bother ,i have some quesions:
- https://github.com/tc39/proposal-source-phase-imports `import source` proposal Is this what you mean? Now TS seems to have no issues in the repository to follow up on this.
- This PR seems to support some other syntaxes like `import.defer("./a.js").then()`、`import defer foo from "a"`、 `import type defer * as ns1 from "a";`, but i only see `import defer * as a from "b"` this syntax in the original proposal: https://github.com/tc39/proposal-defer-import-eval/ , So do all these other syntaxes need to be supported?

nicolo-ribaudo commented:
- Yes, and it's not supported yet by TypeScript: https://www.typescriptlang.org/play/?ts=5.8.0-dev.20250121#code/JYWwDg9gTgLgBAZwgVygYwKZwB5wGZQQhwBEAPAmBmsHsBlAHwkDcQA
- `import.defer("./a.js")` is defined at https://tc39.es/proposal-defer-import-eval/#sec-ecmascript-language-expressions. For the others, this PR is testing that they are disallowed :)

rbuckton commented on src/compiler/utilities.ts:
```diff
@@ -12074,7 +12074,7 @@ export function forEachDynamicImportOrRequireCall<IncludeTypeSpaceImports extend
 ): void {
     const isJavaScriptFile = isInJSFile(file);
-    const r = /import|require/g;
+    const r = /import|defer|require/g;
```
I'm not sure we need to expand this to include `defer` as that will hit more than a few false positives and slow down parsing existing code. As an alternative, we could modify `getNodeAtPosition` (which is only used by this function) to stop and return `current` when `isMetaProperty(child)` is `true`.

nicolo-ribaudo replied:
I first tried to get the parent from the node returned by `getNodeAtPosition`, just to discover that it's not set. I didn't notice that `getNodeAtPosition` is only used here, I'll update it :+1:

rbuckton replied:
> I first tried to get the parent from the node returned by `getNodeAtPosition`, just to discover that it's not set.

program.ts calls this function before bind, so `.parent` is not set. In fact, that callee explicitly calls `setParentRecursively` in the callback as this is an expected case.

rbuckton commented on src/compiler/checker.ts:
```diff
@@ -37598,6 +37606,10 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
 
         if (node.keywordToken === SyntaxKind.ImportKeyword) {
+            if (node.name.escapedText === "defer") {
+                Debug.assert(!isCallExpression(node.parent) || node.parent.expression !== node, "Trying to get the type of `import.defer` in `import.defer(...)`");
```
This assertion seems like it could result in an exception in an editor as you type `import.defer`. I'd rather we not crash here.

nicolo-ribaudo replied:
I don't think this can ever cause a crash. These are two cases:
- we have `import.defer(...)`
- we have `import.defer` outside of a call

In the first case, we are handling it in the `CallExpression` visitor and we do not recourse into the `import.defer` node. This assertions is verifying this.

In the second case, `!isCallExpression(node.parent) || node.parent.expression !== node` is true so the assertion does not fail.

rbuckton commented on src/compiler/checker.ts:
```diff
@@ -49740,7 +49756,13 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
+            try {
+                return getRegularTypeOfExpression(node as Expression);
+            }
+            catch (e) {
+                console.error("Error while getting the type of", isExpressionNode(node), node.kind, (node as MetaProperty).keywordToken !== SyntaxKind.ImportKeyword, (node as MetaProperty).name?.escapedText);
+                throw e;
```
Is this left over from local debugging?

nicolo-ribaudo replied:
Whops :sweat_smile:

nicolo-ribaudo commented:
Note: this proposal reached stage 3 at the last TC39 meeting.

weswigham commented on src/compiler/checker.ts:
```diff
@@ -52864,11 +52891,24 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
+                return grammarErrorOnNode(node, Diagnostics.Named_imports_are_not_allowed_in_a_deferred_import);
+            }
+            if (moduleKind !== ModuleKind.ESNext && moduleKind !== ModuleKind.NodeNext) {
+                return grammarErrorOnNode(node, Diagnostics.Deferred_imports_are_only_supported_when_the_module_flag_is_set_to_esnext_or_nodenext);
```
Even in `nodenext`, they also probably shouldn't be allowed in `cjs`-targeting files, unless we've actually implemented a downlevelling. So this also needs to check the emit module kind for the containing source file.

weswigham commented on src/compiler/types.ts:
```diff
@@ -487,7 +488,7 @@ export const enum SyntaxKind {
     LastJSDocTagNode = JSDocImportTag,
     /** @internal */ FirstContextualKeyword = AbstractKeyword,
-    /** @internal */ LastContextualKeyword = OfKeyword,
+    /** @internal */ LastContextualKeyword = DeferKeyword,
```
```suggestion
    /** @internal */ LastContextualKeyword = LastKeyword,

```
Just to prevent needing to edit it again.

nicolo-ribaudo commented:
@weswigham Actually, reading https://www.typescriptlang.org/docs/handbook/modules/theory.html#module-resolution, should `nodenext` allow this at all?
> - [nodenext](https://www.typescriptlang.org/docs/handbook/modules/reference.html#node16-node18-nodenext): A moving target reflecting the latest Node.js versions as Node.js’s module system evolves. As of TypeScript 5.8, nodenext supports require of ECMAScript modules.
> - [esnext](https://www.typescriptlang.org/docs/handbook/modules/reference.html#es2015-es2020-es2022-esnext): Currently identical to es2022, but will be a moving target reflecting the latest ECMAScript specifications, as well as module-related Stage 3+ proposals that are expected to be included in upcoming specification versions.

The latest Node.js version does not support this proposal yet.

nicolo-ribaudo commented:
@weswigham All your suggestions have been applied, except for the `nodenext` one.

I ended up dropping support for `nodenext`, since according to what I read in the docs I don't think it should support syntax that Node.js does not support yet. I can restore it (fixed) if you disagree. 

I also added support for the new syntax when `module: preserve`.

weswigham reviewed (approved):
We'll need to remember that when we add support for these to `node` module modes to make support format-dependent and/or downlevel it, but for now, this is probably fine.

nicolo-ribaudo commented:
PR coming leter today!
